src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -632,6 +648,57 @@ private static MemberInfo GetMemberInfo<T>(Expression<T> expr)
             return mc.Member;
         }
 
+        // The result of the method is null so we fallback to some runtime logic.
+        // First we check if the result is IResult, Task<IResult> or ValueTask<IResult>. If
+        // it is, we await if necessary then execute the result.
+        // Then we check to see if it's Task<object> or ValueTask<object>. If it is, we await
+        // if necessary and restart the cycle until we've reached a terminal state (unknown type).
+        // We currently don't handle Task<unknown> or ValueTask<unknown>. We can support this later if this
+        // ends up being a common scenario.
+        private static async Task ExecuteObjectReturn(object? obj, HttpContext httpContext)
+        {
+        start:
+            if (obj is IResult result)
+            {
+                await ExecuteResultWriteResponse(result, httpContext);
+            }
+            else if (obj is string stringValue)
+            {
+                await httpContext.Response.WriteAsync(stringValue);
+            }
+            else if (obj is Task<IResult?> task)
+            {
+                await ExecuteTaskResult(task, httpContext);
+            }
+            else if (obj is ValueTask<IResult?> valueTask)
+            {
+                await ExecuteValueTaskResult(valueTask, httpContext);
+            }
+            else if (obj is Task<string?> taskString)
+            {
+                await ExecuteTaskOfString(taskString, httpContext);
+            }
+            else if (obj is ValueTask<string?> valueTaskString)
+            {
+                await ExecuteValueTaskOfString(valueTaskString, httpContext);
+            }
+            else if (obj is Task<object> taskObj)
+            {
+                obj = await taskObj;
+                goto start;
```

Comment:
Okay so I missed this earlier. I'm not sure I like this. I don't think we need it at all.

If we must do this, why not just make it recursive? And are there any test cases for `Task<ValueTask<string>>` and the like.

Reply:
Because gotos are the bomb and this just works! I can add a test case for `Task<VT<string>>` but you'll likely hit compiler limitations with nested generics before you break out codez